tools/pubsys/src/vmware/upload_ova/mod.rs

```diff
@@ -0,0 +1,246 @@
+//! The upload_ova module owns the 'upload_ova' subcommand and is responsible for collating all of
+//! the config necessary to upload an OVA bundle to VMware datacenters.
+use crate::vmware::govc::Govc;
+use crate::Args;
+use log::{debug, info, trace};
+use pubsys_config::vmware::{
+    Datacenter, DatacenterBuilder, DatacenterCreds, DatacenterCredsBuilder, DatacenterCredsConfig,
+    VMWARE_CREDS_PATH,
+};
+use pubsys_config::InfraConfig;
+use serde::Serialize;
+use snafu::{ensure, OptionExt, ResultExt};
+use std::convert::TryInto;
+use std::fs;
+use std::path::PathBuf;
+use structopt::StructOpt;
+use tempfile::NamedTempFile;
+use tinytemplate::TinyTemplate;
+
+const SPEC_TEMPLATE_NAME: &str = "spec_template";
+
+/// Uploads a Bottlerocket OVA to VMware datacenters
+#[derive(Debug, StructOpt)]
+#[structopt(setting = clap::AppSettings::DeriveDisplayOrder)]
+pub(crate) struct UploadArgs {
+    /// Path to the OVA image
+    #[structopt(short = "o", long, parse(from_os_str))]
+    ova: PathBuf,
+
+    /// Path to the import spec
+    #[structopt(short = "s", long, parse(from_os_str))]
+    spec: PathBuf,
+
+    /// The desired VM name
+    #[structopt(short = "n", long)]
+    name: String,
+
+    /// Make the uploaded OVA a VM template
+    #[structopt(long)]
+    mark_as_template: bool,
+
+    /// Datacenters to which you want to upload the OVA
+    #[structopt(long, use_delimiter = true)]
+    datacenters: Vec<String>,
+}
+
+/// Common entrypoint from main()
+pub(crate) fn run(args: &Args, upload_args: &UploadArgs) -> Result<()> {
+    // Get infra config
+    info!(
+        "Checking for infra config at path: {}",
+        args.infra_config_path.display()
+    );
+    let infra_config =
+        InfraConfig::from_path_or_default(&args.infra_config_path).context(error::InfraConfig)?;
+    trace!("Using infra config: {:?}", infra_config);
+
+    let vmware = infra_config
+        .vmware
+        .context(error::MissingConfig { missing: "vmware" })?;
+
+    // If the user gave an override list of datacenters, use it, otherwise use what's in the config
+    let upload_datacenters = if !upload_args.datacenters.is_empty() {
+        &upload_args.datacenters
+    } else {
+        &vmware.datacenters
+    };
+    ensure!(
+        !upload_datacenters.is_empty(),
+        error::MissingConfig {
+            missing: "vmware.datacenters"
+        }
+    );
+
+    // Retrieve credentials from GOVC_ environment variables
+    let creds_env = DatacenterCredsBuilder::from_env();
+    // Retrieve credentials from file. The `home` crate is used to construct the VMWARE_CREDS_PATH,
+    // and it's possible (however unlikely) that it is unable to determine the user;s home folder.
```

Comment:
```suggestion
    // and it's possible (however unlikely) that it is unable to determine the user's home folder.
```

Reply:
That pesky little finger just couldn't reach over one more key... :)